src/main/java/com/google/idea/perf/Autocomplete.kt

```diff
@@ -33,34 +33,48 @@ class Autocomplete {
         val token = tokens.getOrElse(tokenIndex) { "" }
 
         return when (tokenIndex) {
-            0 -> predictToken(
-                listOf("clear", "reset", "trace", "untrace"),
-                token,
-                cancellationCheck
+            0 -> predictOptionToken(
+                listOf("clear", "reset", "trace", "untrace"), token, cancellationCheck
             )
             1 -> when (command) {
-                is TracerCommand.Trace -> predictToken(
-                    listOf("all", "count", "wall-time"),
-                    token,
-                    cancellationCheck
+                is TracerCommand.Trace -> predictOptionToken(
+                    listOf("all", "count", "wall-time"), token, cancellationCheck
                 )
                 else -> emptyList()
             }
             2 -> when (command) {
-                is TracerCommand.Trace -> predictToken(classNames, token, cancellationCheck)
+                is TracerCommand.Trace -> predictClassToken(classNames, token, cancellationCheck)
                 else -> emptyList()
             }
             else -> emptyList()
         }
     }
 
+    private fun predictOptionToken(
+        choices: Collection<String>,
+        token: String,
+        cancellationCheck: () -> Unit
+    ): List<Suggestion> {
+        return predictToken(choices, token, { it.formattedSource }, { null }, cancellationCheck)
+    }
+
+    private fun predictClassToken(
+        choices: Collection<String>,
+        token: String,
+        cancellationCheck: () -> Unit
+    ): List<Suggestion> {
+        return predictToken(choices, token, { it.formattedSource }, { null }, cancellationCheck)
+    }
+
     private fun predictToken(
         choices: Collection<String>,
         token: String,
+        nameFunc: (MatchResult) -> String,
+        detailFunc: (MatchResult) -> String?,
```

Comment:
I'm confused by this layer of indirection via `nameFunc` and `detailFunc`. Will it be useful later on?

(I was also tempted to ask the same thing about the `cancellationCheck` lambda... i.e., I think it's OK to call `ProgressManager.checkCanceled()` directly since that is already the IntelliJ-wide standard.)

Reply:
Yup. `nameFunc` refers to the main text you see in the autocomplete list, and `detailFunc` refers to the gray text seen on the right side of the list (yet to be displayed in the UI).